### PR TITLE
credentials/xds: ServerHandshake() implementation

### PR DESCRIPTION
**Pull request by easwars**
Summary of changes:
- `NewServerCredentials()` which returns a `TransportCredentials` implementation for use on the server-side.
- `ServerHandshake()` implementation which does the following:
  - Expects the passed in `net.Conn` to implement `XDSHandshakeInfo()`, and if so, gets the certificate providers from the returned `HandshakeInfo`. If not, uses `fallback` credentials.
  - Expects the passed in `net.Conn` to implement a `GetDeadline()`, which is used to create the `context` passed to `certprovider.KeyMaterial()`.

The corresponding server-side changes will be sent out in a separate PR since this is big enough and separate enough to be considered on its own.

---
This change is [Reviewable](https://reviewable.io/reviews/grpc/grpc-go/4089)

### CODE REVIEW

**Comment by easwars** [comment]
@dfawley Just FYI, the full server-side security implementation is at: https://github.com/grpc/grpc-go/pull/4092

I realized that I'm not able to send the xDS server implementation separately because it uses the `NewServerCredentials()` function which is implemented here.

File: credentials/xds/xds.go [hunk 1]
```diff
@@ -173,13 +204,14 @@ func (hi *HandshakeInfo) makeTLSConfig(ctx context.Context) (*tls.Config, error)
 	// includes hostname verification) or none. We are forced to go with the
 	// latter and perform the normal cert validation ourselves.
 	cfg := &tls.Config{InsecureSkipVerify: true}
-	if rootProv != nil {
-		km, err := rootProv.KeyMaterial(ctx)
-		if err != nil {
-			return nil, fmt.Errorf("xds: fetching trusted roots from CertificateProvider failed: %v", err)
-		}
-		cfg.RootCAs = km.Roots
+
```
Comment (ZhenLian) [inline credentials/xds/xds.go:207]: Can I know why we don't need to check `rootProv` is nil here? Is it because we always make a call to `validate` and then call `make*SideTLSConfig` here? Would it be better if we merge `validate` and `make*SideTLSConfig` into one single function?

File: credentials/xds/xds.go [hunk 2]
```diff
@@ -129,6 +152,14 @@ func (hi *HandshakeInfo) SetAcceptedSANs(sans []string) {
 	hi.mu.Unlock()
 }
 
+// SetRequireClientCert updates whether a client cert is required during the
+// ServerHandshake(). A value of true indicates that we are performing mTLS.
+func (hi *HandshakeInfo) SetRequireClientCert(require bool) {
```
Comment (ZhenLian) [inline credentials/xds/xds.go:157]: Just curious: what is the motivation behind this change? Does xDS allows one-side TLS and mTLS connections in the same time?

Comment (ZhenLian) [reply credentials/xds/xds.go:157]: Also, it seems this function is only used in tests. Do we expect to expose it as a public function?

**Review by ZhenLian — CHANGES_REQUESTED** [review]
Looks good to me overall! Just some minor comments.

**Review by dfawley — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/grpc/grpc-go/4089)* status: 0 of 3 files reviewed, 4 unresolved discussions (waiting on @dfawley, @easwars, and @ZhenLian)

___
*[credentials/xds/xds.go, line 77 at r1](https://reviewable.io/reviews/grpc/grpc-go/4089#-MO3i4CM-OLWr2NfF5p1:-MO3i4CM-OLWr2NfF5p2:b-7ynwo1) ([raw file](https://github.com/grpc/grpc-go/blob/23a79514f1e96e11ab77b9287240a6b827f28a26/credentials/xds/xds.go#L77)):*
> ```Go
> 	// FallbackCreds specifies the fallback credentials to be used when the
> 	// management server does not return any security configuration. Attempts to
> 	// create server credentials without a fallback credentials will fail.
> ```

"...without fallback credentials..." (remove "a")
___
*[credentials/xds/xds.go, line 123 at r1](https://reviewable.io/reviews/grpc/grpc-go/4089#-MO8L22dESjxJjcK-VOQ:-MO8L22dESjxJjcK-VOR:b-lgas5x) ([raw file](https://github.com/grpc/grpc-go/blob/23a79514f1e96e11ab77b9287240a6b827f28a26/credentials/xds/xds.go#L123)):*
> ```Go
> //
> // Safe for concurrent access.
> type HandshakeInfo struct {
> ```

Can we hide all of the non-user-facing functionality in this file in `internal`?

(*Not* in this PR!)
___
*[credentials/xds/xds.go, line 238 at r1](https://reviewable.io/reviews/grpc/grpc-go/4089#-MODSlpL8znbo9YVE2WZ:-MODSlpL8znbo9YVE2W_:b-3cy4ev) ([raw file](https://github.com/grpc/grpc-go/blob/23a79514f1e96e11ab77b9287240a6b827f28a26/credentials/xds/xds.go#L238)):*
<details><summary><i alt="summary">Quoted 11 lines of code…</i></summary>

> ```
> 	hi.mu.Lock()
> 	// Since the call to KeyMaterial() can block, we read the providers under
> 	// the lock but call the actual function after releasing the lock.
> 	rootProv, idProv := hi.rootProvider, hi.identityProvider
> 	requireClientCert := hi.requireClientCert
> 	hi.mu.Unlock()
> 	clientAuth := tls.NoClientCert
> 	if requireClientCert {
> 		clientAuth = tls.RequireAndVerifyClientCert
> 	}
> 	cfg := &tls.Config{ClientAuth: clientAuth}
> ```
</details>


Optional:

```
cfg := &tls.Config{ClientAuth: tls.NoClientCert}
hi.mu.Lock()
rootProv, idProv := hi.rootProvider, hi.identityProvider
if hi.requireClientCert {
  cfg.ClientAuth = tls.RequireAndVerifyClientCert
}
hi.mu.Unlock()
```

(Remove local var.)
___
*[credentials/xds/xds_client_test.go, line 415 at r1](https://reviewable.io/reviews/grpc/grpc-go/4089#-MODUusjAE-T29KvYlLw:-MODUusjAE-T29KvYlLx:b-b9inqv) ([raw file](https://github.com/grpc/grpc-go/blob/23a79514f1e96e11ab77b9287240a6b827f28a26/credentials/xds/xds_client_test.go#L415)):*
> ```
> 		var b []byte
> 		_, err := rawConn.Read(b)
> ```

I'm not sure I understand this - won't this read 0 bytes and return immediately, because `b` hasn't been allocated?
___
*[credentials/xds/xds_server_test.go, line 258 at r1](https://reviewable.io/reviews/grpc/grpc-go/4089#-MODWwOT01VmKSMWA2ty:-MODWwOU3Gs_F0qLwbQh:b-2lz6xl) ([raw file](https://github.com/grpc/grpc-go/blob/23a79514f1e96e11ab77b9287240a6b827f28a26/credentials/xds/xds_server_test.go#L258)):*
> ```Go
> 	tlsConn := tls.Client(rawConn, makeClientTLSConfig(t, true))
> 	tlsConn.SetDeadline(time.Now().Add(defaultTestTimeout))
> 	// See comment in makeClientTLSConfig() on why we ignore this error.
> ```

This says "Handshake always seems to succeed".  Shouldn't we still catch the error and `Fatal` in case that isn't the case?

File: credentials/xds/xds.go (hunk 2 above)
Comment (easwars) [reply credentials/xds/xds.go:160]: I do not quite understand what you mean by "Does xDS allows one-side TLS and mTLS connections in the same time?". In a single gRPC application, one could be exposing multiple services. And this can be done with a single `xds.GRPCServer` and single `xds.Credentials`. But each of these services could listen on different ports. This means that the listener configuration could be different for them, and hence could be TLS for one and mTLS for another. And since the credentials object is shared between the two services, we need to pass all handshake info on a per-connection basis and we cannot store anything at the credentials level.

This function is not something that will only be used by tests. This will be used by `xds.GRPCServer` to set this field based on what it receives from the control plane.

File: credentials/xds/xds.go (hunk 1 above)
Comment (easwars) [reply credentials/xds/xds.go:196]: Thanks for the suggestion. Incorporated it and got rid of the `validate()` function.

**Review by easwars — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/grpc/grpc-go/4089)* status: 0 of 3 files reviewed, 6 unresolved discussions (waiting on @dfawley and @easwars)

___
*[credentials/xds/xds.go, line 77 at r1](https://reviewable.io/reviews/grpc/grpc-go/4089#-MO3i4CM-OLWr2NfF5p1:-MOE9EZTBiGUvkelIQlt:b-896fix) ([raw file](https://github.com/grpc/grpc-go/blob/23a79514f1e96e11ab77b9287240a6b827f28a26/credentials/xds/xds.go#L77)):*
<details><summary><i>Previously, dfawley (Doug Fawley) wrote…</i></summary><blockquote>

"...without fallback credentials..." (remove "a")
</blockquote></details>

Done.
___
*[credentials/xds/xds.go, line 123 at r1](https://reviewable.io/reviews/grpc/grpc-go/4089#-MO8L22dESjxJjcK-VOQ:-MOE9WZd0ZgYayMQObyQ:b-pmbskr) ([raw file](https://github.com/grpc/grpc-go/blob/23a79514f1e96e11ab77b9287240a6b827f28a26/credentials/xds/xds.go#L123)):*
<details><summary><i>Previously, dfawley (Doug Fawley) wrote…</i></summary><blockquote>

Can we hide all of the non-user-facing functionality in this file in `internal`?

(*Not* in this PR!)
</blockquote></details>

Sure, added a TODO for the same.
___
*[credentials/xds/xds.go, line 238 at r1](https://reviewable.io/reviews/grpc/grpc-go/4089#-MODSlpL8znbo9YVE2WZ:-MOEAeys7iY4h1z49NP2:bs7wumb) ([raw file](https://github.com/grpc/grpc-go/blob/23a79514f1e96e11ab77b9287240a6b827f28a26/credentials/xds/xds.go#L238)):*
<details><summary><i>Previously, dfawley (Doug Fawley) wrote…</i></summary><blockquote>

> ```
> 	hi.mu.Lock()
> 	// Since the call to KeyMaterial() can block, we read the providers under
> 	// the lock but call the actual function after releasing the lock.
> 	rootProv, idProv := hi.rootProvider, hi.identityProvider
> 	requireClientCert := hi.requireClientCert
> 	hi.mu.Unlock()
> 	clientAuth := tls.NoClientCert
> 	if requireClientCert {
> 		clientAuth = tls.RequireAndVerifyClientCert
> 	}
> 	cfg := &tls.Config{ClientAuth: clientAuth}
> ```

Optional:

```
cfg := &tls.Config{ClientAuth: tls.NoClientCert}
hi.mu.Lock()
rootProv, idProv := hi.rootProvider, hi.identityProvider
if hi.requireClientCert {
  cfg.ClientAuth = tls.RequireAndVerifyClientCert
}
hi.mu.Unlock()
```

(Remove local var.)
</blockquote></details>

Done.
___
*[credentials/xds/xds_client_test.go, line 415 at r1](https://reviewable.io/reviews/grpc/grpc-go/4089#-MODUusjAE-T29KvYlLw:-MOEG5L336HdBS1DZobo:b-wpyygu) ([raw file](https://github.com/grpc/grpc-go/blob/23a79514f1e96e11ab77b9287240a6b827f28a26/credentials/xds/xds_client_test.go#L415)):*
<details><summary><i>Previously, dfawley (Doug Fawley) wrote…</i></summary><blockquote>

> ```
> 		var b []byte
> 		_, err := rawConn.Read(b)
> ```

I'm not sure I understand this - won't this read 0 bytes and return immediately, because `b` hasn't been allocated?
</blockquote></details>

oops. Didn't realize that passing a nil slice would cause `Read()` to return immediately.
Changed the server side handshake function to simply block on a channel which is closed only after the client side handshake times out.
Also added code to check that the error returned by the handshake function is the one we expect, so that we don't inadvertently accept other errors.
___
*[credentials/xds/xds_server_test.go, line 258 at r1](https://reviewable.io/reviews/grpc/grpc-go/4089#-MODWwOT01VmKSMWA2ty:-MOEHG2xEUaN7_Hm8BSE:b-896fix) ([raw file](https://github.com/grpc/grpc-go/blob/23a79514f1e96e11ab77b9287240a6b827f28a26/credentials/xds/xds_server_test.go#L258)):*
<details><summary><i>Previously, dfawley (Doug Fawley) wrote…</i></summary><blockquote>

This says "Handshake always seems to succeed".  Shouldn't we still catch the error and `Fatal` in case that isn't the case?
</blockquote></details>

Done.

**Review by dfawley — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/grpc/grpc-go/4089)* status: 0 of 3 files reviewed, 6 unresolved discussions (waiting on @dfawley, @easwars, and @ZhenLian)

___
*[credentials/xds/xds.go, line 123 at r1](https://reviewable.io/reviews/grpc/grpc-go/4089#-MO8L22dESjxJjcK-VOQ:-MOHrJ1M60IPThBAXnlW:boupuhq) ([raw file](https://github.com/grpc/grpc-go/blob/23a79514f1e96e11ab77b9287240a6b827f28a26/credentials/xds/xds.go#L123)):*
<details><summary><i>Previously, easwars (Easwar Swaminathan) wrote…</i></summary><blockquote>

Sure, added a TODO for the same.
</blockquote></details>

Please file an issue, too, unless you plan to do this ASAP.

**Comment by easwars** [comment]
> > _[Reviewable](https://reviewable.io/reviews/grpc/grpc-go/4089)_ status: 0 of 3 files reviewed, 6 unresolved discussions (waiting on @dfawley, @easwars, and @ZhenLian)
> 
> _[credentials/xds/xds.go, line 123 at r1](https://reviewable.io/reviews/grpc/grpc-go/4089#-MO8L22dESjxJjcK-VOQ:-MOHrJ1M60IPThBAXnlW:boupuhq) ([raw file](https://github.com/grpc/grpc-go/blob/23a79514f1e96e11ab77b9287240a6b827f28a26/credentials/xds/xds.go#L123)):_
> 
> Previously, easwars (Easwar Swaminathan) wrote…
> Please file an issue, too, unless you plan to do this ASAP.

Done. Thanks.
https://github.com/grpc/grpc-go/issues/4100

File: credentials/xds/xds.go (hunk 2 above)
Comment (ZhenLian) [reply credentials/xds/xds.go:160]: OK I see. I think your explanation explains it well :)

**Review by ZhenLian — APPROVED** [review]
> Reviewed 2 of 3 files at r3, 1 of 2 files at r4.
> *[Reviewable](https://reviewable.io/reviews/grpc/grpc-go/4089)* status: 3 of 4 files reviewed, 5 unresolved discussions (waiting on @dfawley and @ZhenLian)
